tests/unauthenticated_stream.rs

```diff
@@ -0,0 +1,264 @@
+                assert_eq!(Err(error::Unspecified), o_result);
+            },
+            Some(error) => {
+                unreachable!("Unexpected error test case: {}", error);
+            },
+        };
+
+        Ok(())
+    });
+}
+
+fn test_stream_key_sizes(stream_alg: &'static unauthenticated_stream::Algorithm) {
+    let key_len = stream_alg.key_len();
+    let key_data = vec![0u8; key_len * 2];
+
+    // Key is the right size.
+    assert!(unauthenticated_stream::DecryptingKey::new(stream_alg,
+        &key_data[..key_len]).is_ok());
+    assert!(unauthenticated_stream::EncryptingKey::new(stream_alg,
+        &key_data[..key_len]).is_ok());
+
+    // Key is one byte too small.
+    assert!(unauthenticated_stream::DecryptingKey::new(stream_alg,
+        &key_data[..(key_len - 1)]).is_err());
+    assert!(unauthenticated_stream::EncryptingKey::new(stream_alg,
+        &key_data[..(key_len - 1)]).is_err());
+
+    // Key is one byte too large.
+    assert!(unauthenticated_stream::DecryptingKey::new(stream_alg,
+        &key_data[..(key_len + 1)]).is_err());
+    assert!(unauthenticated_stream::EncryptingKey::new(stream_alg,
+        &key_data[..(key_len + 1)]).is_err());
+
+    // Key is half the required size.
+    assert!(unauthenticated_stream::DecryptingKey::new(stream_alg,
+        &key_data[..(key_len / 2)]).is_err());
+    assert!(unauthenticated_stream::EncryptingKey::new(stream_alg,
+        &key_data[..(key_len / 2)]).is_err());
+
+    // Key is twice the required size.
+    assert!(unauthenticated_stream::DecryptingKey::new(stream_alg,
+        &key_data[..(key_len * 2)]).is_err());
+    assert!(unauthenticated_stream::EncryptingKey::new(stream_alg,
+        &key_data[..(key_len * 2)]).is_err());
+
+    // Key is empty.
+    assert!(unauthenticated_stream::DecryptingKey::new(stream_alg,
+        &[]).is_err());
+    assert!(unauthenticated_stream::EncryptingKey::new(stream_alg,
+        &[]).is_err());
+
+    // Key is one byte.
+    assert!(unauthenticated_stream::DecryptingKey::new(stream_alg,
+        &[0]).is_err());
+    assert!(unauthenticated_stream::EncryptingKey::new(stream_alg,
+        &[0]).is_err());
+}
+
+// Test that we reject non-standard nonce sizes.
+//
+// XXX: This test isn't that great in terms of how it tests
+// `open_in_place`. It should be constructing a valid ciphertext using the
+// unsupported nonce size using a different implementation that supports
+// non-standard nonce sizes. So, when `open_in_place` returns
+// `Err(error::Unspecified)`, we don't know if it is because it rejected
+// the non-standard nonce size or because it tried to process the input
+// with the wrong nonce. But at least we're verifying that `open_in_place`
+// won't crash or access out-of-bounds memory (when run under valgrind or
+// similar). The AES-128-GCM tests have some WRONG_NONCE_LENGTH test cases
+// that tests this more correctly.
+fn test_stream_nonce_sizes(stream_alg: &'static unauthenticated_stream::Algorithm)
```

Comment:
I guess this test is exactly the same as the test for the AEAD code, except that unauthenticated. Let's find a way to factor out the common logic so that it can be shared for the AEAD and unauthenticated tests.

Reply:
I tried to go the macro way like for the other test, but the problem here is that since the aead API also needs to pass tag and ad around, the code ends up being quite different from the unathenticated encryption case, so there's not much code reuse. But maybe there's a way to do this that I missed.